### PR TITLE
1 initial

### PR DESCRIPTION
**Pull request by loadzero**
Initial kinesis integration work and test driver.

### CODE REVIEW

File: kinesis_consumer.cpp [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+// create the state for the consumer.
+// note - this is a blocking call wrt obtaining the shard iterator.
+// on error this returns NULL.
+kinesis_consumer*
+kinesis_consumer_create()
+{
+	Aws::Utils::Logging::InitializeAWSLogging(Aws::MakeShared<AltLogSystem>("logging", Aws::Utils::Logging::LogLevel::Error));
+
+	ClientConfiguration config;
+	config.region = Aws::Region::US_WEST_2;
```
Comment (derekjn) [inline kinesis_consumer.cpp:84]: All of the configuration in this function should be derived from arguments.

File: build_aws.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
```
Comment (derekjn) [inline build_aws.sh]: Let's just put these dependency build steps in the `README` like [pipeline_kafka](https://github.com/pipelinedb/pipeline_kafka) does for `librdkafka`.

File: kinesis_consumer.cpp [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+	KinesisClient &client = *kc->client;
+
+	double last_request_time = 0.0;
+
+	GetRecordsRequest req;
+
+	while (kc->keep_running)
+	{
+		req.SetShardIterator(kc->shard_iter);
+		req.SetLimit(1000);
```
Comment (derekjn) [inline kinesis_consumer.cpp:153]: This limit should probably come from a member of the `kinesis_consumer`, even if it's hard-coded as `1000` for now when the consumer is created.

File: kinesis_consumer.cpp [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+	config.region = Aws::Region::US_WEST_2;
+	config.verifySSL = false;
+	
+	config.retryStrategy = Aws::MakeShared<DefaultRetryStrategy>("kinesis_consumer", 5, 25);
+	auto *kc = new KinesisClient(config);
+
+	GetShardIteratorRequest request;
+
+	request.SetStreamName("test");
+	request.SetShardId("shardId-000000000000");
```
Comment (derekjn) [inline kinesis_consumer.cpp:93]: This needs to know how to automatically consume from all shards, as in `pipeline_kafka`:

https://github.com/pipelinedb/pipeline_kafka/blob/master/pipeline_kafka.c#L714

File: kinesis_consumer.cpp [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+using namespace Aws::Kinesis::Model;
+using namespace Aws::Utils;
+
+struct kinesis_consumer
+{
+	KinesisClient *client;
+	concurrent_queue<GetRecordsOutcome*> *queue;
+	std::atomic<bool> keep_running;
+	std::thread *thread;
+	Aws::String shard_iter;
```
Comment (derekjn) [inline kinesis_consumer.cpp:48]: Each consumer is probably going to need some kind of list of shards that it is responsible for reading from.

File: kinesis_consumer.cpp (hunk 1 above)
Comment (loadzero) [inline kinesis_consumer.cpp:84]: Concur, leaving hardcoded until I'm finished with the next layer up.

File: build_aws.sh (hunk 2 above)
Comment (loadzero) [inline build_aws.sh]: done

File: kinesis_consumer.cpp (hunk 3 above)
Comment (loadzero) [inline kinesis_consumer.cpp:153]: Yeah, there does need to be some fiddle factor in there. This is the KCL default. It is complicated by this (from https://docs.aws.amazon.com/kinesis/latest/APIReference/API_GetRecords.html):

So, some sort of backoff / adaptation may be required.

```
Each data record can be up to 1 MB in size, and each shard can read up to 2 MB 
per second. You can ensure that your calls don't exceed the maximum supported 
size or throughput by using the Limit parameter to specify the maximum number 
of records that GetRecords can return. Consider your average record size when 
determining this limit.

The size of the data returned by GetRecords varies depending on the utilization 
of the shard. The maximum size of data that GetRecords can return is 10 MB. If 
a call returns this amount of data, subsequent calls made within the next 5 
seconds throw ProvisionedThroughputExceededException. If there is insufficient 
provisioned throughput on the shard, subsequent calls made within the next 1 
second throw ProvisionedThroughputExceededException. Note that GetRecords won't 
return any data when it throws an exception. For this reason, we recommend that 
you wait one second between calls to GetRecords; however, it's possible that 
the application will get exceptions for longer than 1 second.
```

File: kinesis_consumer.cpp (hunk 4 above)
Comment (loadzero) [inline kinesis_consumer.cpp:93]: Next PR.

File: kinesis_consumer.cpp (hunk 5 above)
Comment (loadzero) [inline kinesis_consumer.cpp:48]: Next PR.

**Pull request closed without merging**